File: lifewatch/server/services/sync_service.py
"""
数据同步服务
负责从 ActivityWatch 同步数据并分类
"""

import time
from datetime import datetime
from typing import Dict
from lifewatch.server.services.data_processing_service import DataProcessingService


class SyncService:
    """
    数据同步服务
    
    整合现有业务逻辑，实现从 ActivityWatch 同步数据的完整流程
    """
    
    def __init__(self):
        self.data_processor = DataProcessingService()
    
    def sync_from_activitywatch(
        self,
        auto_classify: bool = True
    ) -> Dict:
        """
        增量同步 ActivityWatch 数据（从数据库最新时间同步到现在）
        
        Args:
            auto_classify: 是否自动分类新应用
            
        Returns:
            Dict: 同步结果
        """
        start_time = time.time()
        
        try:
            # 使用 DataProcessingService 处理增量同步
            result = self.data_processor.process_activitywatch_data(
                auto_classify=auto_classify
            )
            
            duration = time.time() - start_time
            
            return {
                "status": "success",
                "synced_events": result["saved_events"],
                "new_apps_classified": result["classified_apps"],
                "duration": round(duration, 2),
                "message": f"成功同步数据（增量模式）",
                "details": {
                    "sync_mode": result["sync_mode"],
                    "time_range": result["time_range"],
                    "total_events": result["total_events"],
                    "filtered_events": result["filtered_events"],
                    "apps_to_classify": result["apps_to_classify"],
                    "unclassified_events": result["unclassified_events"]
                }
            }
            
        except Exception as e:
            duration = time.time() - start_time
            return {
                "status": "failed",
                "synced_events": 0,
                "new_apps_classified": 0,
                "duration": round(duration, 2),
                "message": f"同步失败: {str(e)}"
            }
    
    def sync_by_time_range(
        self,
        start_time: str,
        end_time: str,
        auto_classify: bool = True
    ) -> Dict:
        """
        按时间范围同步 ActivityWatch 数据
        
        Args:
            start_time: 开始时间，格式: YYYY-MM-DD HH:MM:SS
            end_time: 结束时间，格式: YYYY-MM-DD HH:MM:SS
            auto_classify: 是否自动分类新应用
            
        Returns:
            Dict: 同步结果
        """
        sync_start = time.time()
        
        try:
            # 解析时间字符串
            start_dt = datetime.strptime(start_time, "%Y-%m-%d %H:%M:%S")
            end_dt = datetime.strptime(end_time, "%Y-%m-%d %H:%M:%S")
            
            # 使用 DataProcessingService 处理数据
            result = self.data_processor.process_activitywatch_data_by_time_range(
                start_time=start_dt,
                end_time=end_dt,
                auto_classify=auto_classify
            )
            
            duration = time.time() - sync_start
            
            return {
                "status": "success",
                "synced_events": result["saved_events"],
                "new_apps_classified": result["classified_apps"],
                "duration": round(duration, 2),
                "message": f"成功同步时间范围数据",
                "details": {
                    "sync_mode": "time_range",
                    "time_range": result["time_range"],
                    "total_events": result["total_events"],
                    "filtered_events": result["filtered_events"],
                    "apps_to_classify": result["apps_to_classify"],
                    "unclassified_events": result["unclassified_events"]
                }
            }
            
        except Exception as e:
            duration = time.time() - sync_start
            return {
                "status": "failed",
                "synced_events": 0,
                "new_apps_classified": 0,
                "duration": round(duration, 2),
                "message": f"时间范围同步失败: {str(e)}"
            }

File: services/syncService.ts
/**
 * 数据同步服务
 * 负责与后端同步 ActivityWatch 数据
 */

const API_BASE_URL = 'http://localhost:8000/api/v2';

export interface SyncRequest {
    auto_classify?: boolean;
}

export interface SyncTimeRangeRequest {
    start_time: string; // Format: YYYY-MM-DD HH:MM:SS
    end_time: string;   // Format: YYYY-MM-DD HH:MM:SS
    auto_classify?: boolean;
}

export interface SyncResponse {
    status: string;
    synced_events: number;
    new_apps_classified: number;
    duration: number;
    message: string;
    details?: {
        sync_mode: string;
        time_range: string;
        total_events: number;
        filtered_events: number;
        apps_to_classify: number;
        unclassified_events: number;
    };
}

/**
 * 执行增量同步（从数据库最新时间开始同步到现在）
 * @param autoClassify 是否自动分类新应用，默认开启
 */
export async function incrementalSync(autoClassify: boolean = true): Promise<SyncResponse> {
    try {
        const response = await fetch(`${API_BASE_URL}/sync/activitywatch`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                auto_classify: autoClassify,
            }),
        });

        if (!response.ok) {
            throw new Error(`同步失败: ${response.statusText}`);
        }

        const data: SyncResponse = await response.json();
        return data;
    } catch (error) {
        console.error('增量同步错误:', error);
        throw error;
    }
}

/**
 * 按时间范围同步数据
 * @param request 时间范围同步请求参数
 * @returns 同步结果
 */
export async function syncActivityWatchDataByTimeRange(
    request: SyncTimeRangeRequest
): Promise<SyncResponse> {
    try {
        const response = await fetch(`${API_BASE_URL}/sync/activitywatch/timerange`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                start_time: request.start_time,
                end_time: request.end_time,
                auto_classify: request.auto_classify ?? true,
            }),
        });

        if (!response.ok) {
            throw new Error(`同步失败: ${response.statusText}`);
        }

        const data: SyncResponse = await response.json();
        return data;
    } catch (error) {
        console.error('时间范围数据同步错误:', error);
        throw error;
    }
}

